File: bill_tracker/bill_tracker/report/monthly_bill_matrix/monthly_bill_matrix.js
// Global Quick Edit Modal Launcher
window.open_bill_payment_dialog = function (docname) {
    if (!docname) return;

    frappe.db.get_doc("Bill Payment", docname).then(doc => {
        let d = new frappe.ui.Dialog({
            title: __('Update Bill Payment: {0}', [docname]),
            fields: [
                {
                    label: 'Status',
                    fieldname: 'status',
                    fieldtype: 'Select',
                    options: ['Pending', 'Paid'],
                    default: doc.status,
                    reqd: 1
                },
                {
                    label: 'Amount',
                    fieldname: 'amount',
                    fieldtype: 'Currency',
                    default: doc.amount,
                    reqd: 1
                },
                {
                    label: 'Paid Date',
                    fieldname: 'paid_date',
                    fieldtype: 'Date',
                    default: doc.paid_date || frappe.datetime.get_today(),
                    depends_on: 'eval:doc.status=="Paid"'
                }
            ],
            primary_action_label: __('Update Record'),
            primary_action(values) {
                frappe.call({
                    method: 'frappe.client.set_value',
                    args: {
                        doctype: 'Bill Payment',
                        name: docname,
                        fieldname: values
                    },
                    freeze: true,
                    freeze_message: __('Saving...'),
                    callback: function (r) {
                        d.hide();
                        frappe.show_alert({ message: __('Payment updated successfully'), indicator: 'green' });
                        frappe.query_report.refresh();
                    }
                });
            }
        });
        d.show();
    });
};

frappe.query_reports["Monthly Bill Matrix"] = {
    "filters": [],
    "onload": function(report) {
        // Add "Generate Monthly Bills" button to report page header
        report.page.add_inner_button(__('Generate Monthly Bills'), function() {
            frappe.prompt(
                [
                    {
                        label: 'Billing Month',
                        fieldname: 'billing_month',
                        fieldtype: 'Select',
                        options: [
                            'Jan-2026', 'Feb-2026', 'Mar-2026', 'Apr-2026', 
                            'May-2026', 'Jun-2026', 'Jul-2026', 'Aug-2026', 
                            'Sep-2026', 'Oct-2026', 'Nov-2026', 'Dec-2026'
                        ],
                        default: 'Sep-2026',
                        reqd: 1
                    }
                ],
                function(values) {
                    frappe.call({
                        method: 'bill_tracker.api.generate_monthly_bills',
                        args: {
                            billing_month: values.billing_month
                        },
                        freeze: true,
                        freeze_message: __('Generating monthly bills...'),
                        callback: function(r) {
                            if (r.message) {
                                frappe.msgprint(r.message);
                                frappe.query_report.refresh();
                            }
                        }
                    });
                },
                __('Select Billing Month'),
                __('Generate')
            );
        });
    },
    "formatter": function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (!data) return value;

        // Summary Rows Formatting
        if (data.is_summary) {
            if (column.fieldname === "category") {
                return `<b style="color: #111827; font-weight: 700;">${data.category || ''}</b>`;
            }
            if (column.fieldname && column.fieldname.startsWith("amount_")) {
                return `<b>${value}</b>`;
            }
        }

        // Category Status Badges
        if (column.fieldname === "category" && !data.is_summary) {
            if (data.overall_status === "Paid") {
                return `<span style="background-color: #d4edda; color: #155724; padding: 4px 8px; border-radius: 4px; font-weight: 600; display: inline-block;">${data.category}</span>`;
            } else if (data.overall_status === "Pending") {
                return `<span style="background-color: #fff3cd; color: #856404; padding: 4px 8px; border-radius: 4px; font-weight: 600; display: inline-block;">${data.category}</span>`;
            }
        }

        // Interactive Clickable Status Badges
        if (column.fieldname && column.fieldname.startsWith("status_")) {
            let m_key = column.fieldname.replace("status_", "");
            let docName = data[`doc_${m_key}`];
            let statusVal = data[column.fieldname];

            if (docName && statusVal !== "-") {
                let bg = statusVal === "Paid" ? "#d4edda" : "#fff3cd";
                let color = statusVal === "Paid" ? "#155724" : "#856404";

                return `<span onclick="open_bill_payment_dialog('${docName}')" 
                              style="background-color: ${bg}; color: ${color}; padding: 4px 8px; border-radius: 4px; font-weight: 600; display: block; text-align: center; cursor: pointer;"
                              title="Click to quick edit payment">
                              ${statusVal} ✏️
                        </span>`;
            }
        }

        return value;
    },
    "get_datatable_options": function (options) {
        return Object.assign(options, {
            checkboxColumn: false,
            cellHeight: 40,
        });
    }
};

File: bill_tracker/bill_tracker/report/monthly_bill_matrix/monthly_bill_matrix.py
import frappe
from datetime import datetime

def execute(filters=None):
    columns = [
        {"label": "Categories", "fieldname": "category", "fieldtype": "Data", "width": 180},
        {"label": "Bill #", "fieldname": "bill_number", "fieldtype": "Data", "width": 130},
    ]

    raw_months = frappe.get_all("Bill Payment", fields=["billing_month"], distinct=True)
    months_list = [m.billing_month for m in raw_months if m.billing_month]

    def parse_month(m_str):
        try:
            return datetime.strptime(m_str, "%b-%Y")
        except Exception:
            return datetime.min

    months = sorted(months_list, key=parse_month)

    for month in months:
        m_key = month.replace("-", "_").lower()
        columns.extend([
            {"label": f"<b>{month}</b>", "fieldname": f"{m_key}", "fieldtype": "Data", "width": 150},
            # {"label": f"<b>{month}</b><br><span style='font-size: 11px; color: #6b7280;'>Paid Date</span>", "fieldname": f"date_{m_key}", "fieldtype": "Date", "width": 110},
            # {"label": f"<b>{month}</b><br><span style='font-size: 11px; color: #6b7280;'>Status</span>", "fieldname": f"status_{m_key}", "fieldtype": "Data", "width": 100},
        ])

    categories = frappe.get_all("Bill Category", fields=["name", "bill_number"])
    payments = frappe.get_all("Bill Payment", fields=["name", "bill_category", "billing_month", "amount", "due_date", "paid_date", "status"])

    pay_map = {(p.bill_category, p.billing_month): p for p in payments}

    data = []
    pending_totals = {m: 0 for m in months}
    paid_totals = {m: 0 for m in months}

    for cat in categories:
        row = {
            "category": cat.name,
            "bill_number": cat.bill_number or "-",
            "is_summary": False
        }

        has_pending = False
        has_entries = False

        for month in months:
            m_key = month.replace("-", "_").lower()
            p = pay_map.get((cat.name, month))
            if p:
                has_entries = True
                row[f"doc_{m_key}"] = p.name  # Pass payment doc name for quick edit
                row[f"{m_key}"] = p.amount
                row[f"date_{m_key}"] = p.paid_date or p.due_date

                if p.status == "Paid":
                    # row[f"{m_key}"] = p.paid_date or p.due_date
                    paid_totals[month] += (p.amount or 0)
                else:
                    # row[f"{m_key}"] = None
                    pending_totals[month] += (p.amount or 0)
                    has_pending = True

                row[f"status_{m_key}"] = p.status
            else:
                # row[f"amount_{m_key}"] = 0
                # row[f"date_{m_key}"] = None
                row[f"status_{m_key}"] = "-"

        if has_pending:
            row["overall_status"] = "Pending"
        elif has_entries:
            row["overall_status"] = "Paid"
        else:
            row["overall_status"] = "None"

        data.append(row)

    if months:
        # data.append({"is_summary": False})  # Empty row for spacing
        row_pending = {"category": "Amount to Paid", "is_summary": False}
        row_paid = {"category": "Already Paid Amount", "is_summary": False}

        for month in months:
            m_key = month.replace("-", "_").lower()
            row_pending[f"amount_{m_key}"] = pending_totals[month]
            row_paid[f"amount_{m_key}"] = paid_totals[month]

        # data.append(row_pending)
        # data.append(row_paid)

    return columns, data